test/gateway/subscription.js

```diff
@@ -0,0 +1,272 @@
+            url: `http://localhost:${messageServicePort}/graphql`,
+            wsUrl: `ws://localhost:${messageServicePort}/graphql`
+          }]
+        }
+      })
+
+      gateway.listen(0, () => {
+        const ws = new WebSocket(`ws://localhost:${(gateway.server.address()).port}/graphql`, 'graphql-ws')
+        const client = WebSocket.createWebSocketStream(ws, { encoding: 'utf8', objectMode: true })
+        t.tearDown(() => {
+          client.destroy()
+          messageService.close()
+          userService.close()
+          gateway.close()
+          process.exit(0)
+        })
+        client.setEncoding('utf8')
+
+        client.write(JSON.stringify({
+          type: 'connection_init'
+        }))
+
+        client.write(JSON.stringify({
+          id: 1,
+          type: 'start',
+          payload: {
+            query: `
+              subscription {
+                newMessage(toUser: "u1") {
+                  id
+                  text
+                  from {
+                    id
+                    name
+                  }
+                  to {
+                    id
+                    name
+                  }
+                }
+              }
+            `
+          }
+        }))
+
+        client.write(JSON.stringify({
+          id: 2,
+          type: 'start',
+          payload: {
+            query: `
+              subscription {
+                newMessage(toUser: "u2") {
+                  id
+                }
+              }
+            `
+          }
+        }))
+
+        client.write(JSON.stringify({
+          id: 2,
+          type: 'stop'
+        }))
+
+        client.on('data', (chunk) => {
+          const data = JSON.parse(chunk)
+
+          if (data.id === 1 && data.type === 'data') {
+            t.equal(chunk, JSON.stringify({
+              type: 'data',
+              id: 1,
+              payload: {
+                data: {
+                  newMessage: {
+                    id: '1',
+                    text: 'Hi there',
+                    from: {
+                      id: 'u2',
+                      name: 'Jane'
+                    },
+                    to: {
+                      id: 'u1',
+                      name: 'John'
+                    }
+                  }
+                }
+              }
+            }))
+
+            client.end()
+            t.end()
+          } else if (data.id === 2 && data.type === 'complete') {
+            gateway.inject({
+              method: 'POST',
+              url: '/graphql',
+              body: {
+                query: `
+                  mutation {
+                    sendMessage(message: {
+                      text: "Hi there",
+                      fromUserId: "u2",
+                      toUserId: "u1"
+                    }) {
+                      id
+                    }
+                  }
+                `
+              }
+            })
+          }
+        })
+      })
+    })
+  })
+})
```

Comment:
This is way too much nesting for my taste, please linearize it a bit by using an async function.

Reply:
I have a version with awaits but I can't make it work somehow.